hw/mcu/nxp/kinetis/src/hal_spi.c

```diff
@@ -0,0 +1,623 @@
+        master->config.ctarConfig.direction =
+            (settings->data_order == HAL_SPI_MSB_FIRST) ?
+            kDSPI_MsbFirst :
+            kDSPI_LsbFirst;
+        master->config.ctarConfig.bitsPerFrame =
+            (settings->word_size == HAL_SPI_WORD_SIZE_8BIT) ? 8 : 9;
+        master->config.ctarConfig.cpol = cpol;
+        master->config.ctarConfig.cpha = cpha;
+    } else {
+        slave = (struct nxp_spi_slave *) spi;
+        slave->config.ctarConfig.bitsPerFrame =
+            (settings->word_size == HAL_SPI_WORD_SIZE_8BIT) ? 8 : 9;
+        slave->config.ctarConfig.cpol = cpol;
+        slave->config.ctarConfig.cpha = cpha;
+    }
+    return 0;
+}
+
+int
+hal_spi_enable(int spi_num)
+{
+    struct nxp_hal_spi *spi;
+    struct nxp_spi_master *master;
+    struct nxp_spi_slave *slave;
+    spi = hal_spi_resolve(spi_num);
+    if (!spi) {
+        return EINVAL;
+    }
+    if (spi->enabled) {
+        return 0;
+    }
+
+    if (spi->type == HAL_SPI_TYPE_MASTER) {
+        master = (struct nxp_spi_master *) spi;
+        DSPI_MasterInit(spi->dev,
+                        &master->config,
+                        CLOCK_GetFreq(kCLOCK_BusClk));
+    } else {
+        slave = (struct nxp_spi_slave *) spi;
+        DSPI_SlaveInit(spi->dev, &slave->config);
+    }
+
+    spi->enabled = true;
+    NVIC_ClearPendingIRQ(spi->irqn);
+    NVIC_SetVector(spi->irqn, (uint32_t) spi->irq_handler);
+    NVIC_EnableIRQ(spi->irqn);
+    return 0;
+}
+
+int
+hal_spi_disable(int spi_num)
+{
+    struct nxp_hal_spi *spi;
+    spi = hal_spi_resolve(spi_num);
+    if (!spi) {
+        return EINVAL;
+    }
+    if (!spi->enabled) {
+        return 0;
+    }
+
+    DSPI_Deinit(spi->dev);
+
+    spi->enabled = false;
+    NVIC_ClearPendingIRQ(spi->irqn);
+    NVIC_DisableIRQ(spi->irqn);
+    return 0;
+}
+
+uint16_t
+hal_spi_tx_val(int spi_num, uint16_t val)
+{
+    struct nxp_hal_spi *spi;
+    dspi_transfer_t xfer;
+    uint16_t retval = 0;
+    spi = hal_spi_resolve(spi_num);
+    if (!spi) {
+        return EINVAL;
+    }
+
+    if (spi->type == HAL_SPI_TYPE_MASTER) {
+        xfer.txData = (uint8_t *) &val;
+        xfer.rxData = (uint8_t *) &retval;
+        xfer.dataSize = 1;
+        xfer.configFlags = kDSPI_MasterCtar0;
+        DSPI_MasterTransferBlocking(spi->dev, &xfer);
+        return retval;
+    }
+    return 0xFFFF; /* Invalid API. */
+}
+
+int
+hal_spi_txrx(int spi_num, void *txbuf, void *rxbuf, int len)
+{
+    struct nxp_hal_spi *spi;
+    dspi_transfer_t xfer;
+    status_t rc;
+    spi = hal_spi_resolve(spi_num);
+    if (!spi) {
+        return EINVAL;
+    }
+
+    if (spi->type == HAL_SPI_TYPE_MASTER) {
+        xfer.txData = (uint8_t *) txbuf;
+        xfer.rxData = (uint8_t *) rxbuf;
+        xfer.dataSize = len;
+        xfer.configFlags = kDSPI_MasterCtar0;
+        rc = DSPI_MasterTransferBlocking(spi->dev, &xfer);
+        return (rc == kStatus_Success) ? 0 : rc;
```

Comment:
If you initialize `rc` to `-1` you can move this `return` after the `if` block.

Reply:
Addressed, good catch!